### DISCUSSION TITLE
RFC: Command Line DX

### DISCUSSION THREAD

**Discussion by ericvicenti** [rise-tools/rise-tools#153] [Design Doc]
We want to improve the experience for new developers on the TS server. This proposal is inspired by other leaders in the JS and React Native space, namely Expo CLI.

Huge thanks to @arpitBhalla for stepping up on this project with a prototype of [rise start](https://github.com/arpitBhalla/rise-tools/tree/main/packages/rise-start#readme) and [demo-cli](https://github.com/arpitBhalla/rise-tools/tree/main/example/demo-cli)

## Create Rise Project

Create a new rise project:

`npx create rise@next`

This will prompt you for a folder name. And it would be powered by `create-rise` npm package.

The new package will contain a `models` folder and the appropriate NPM dependencies. One npm package contains the new CLI.

## Rise CLI

The new project will contain `@rise-tools/cli` as a `devDependency`

It includes the following commands:

### `npx rise start`

Start a local server that will automatically expose all the models defined in the `models` directory, by path name.

We could include the following arguments:

- `--port=8080` - Specify the port that will be served on. Default is to automatically find an available port
- `--host=localhost` - Specify the host that will be served on
- `--prod` - Start in production mode, which will not watch the models folder for changes. (There may be other differences in behavior for production mode)

When starting in dev mode, a QRCode will be printed in the terminal that allows users to scan with their phone to open the app in Rise Playground.

### `npx rise eject`

This will create a local server and allow users to avoid use of the CLI in production. It can only be run once.

- Create a local `server.ts` file which defines an object of the appropriate models
- Install other necessary dependencies for dev+running without the CLI
- Change the `start` script of the repo 

Eject may have other considerations, for example maybe uninstalling the CLI, and requiring a clean git state so that it will be easy to use git to see what the eject step has accomplished.

**Comment by grabbou** [comment]
This is great, thanks for writing this up and thanks @arpitBhalla for spending your time trying to figure this out!

I would add to this - although it's not super stable at the time of writing - a flag to choose protocols (in case you wanted WebSockets only).

**Comment by grabbou** [comment]
Another thing that comes to my mind is `eject` mechanism. Historically, Expo used to `eject` managed apps and ended up migrating to a better workflow.

What if I wanted to continue to take advantage from the dev experience provided by the CLI, but deploy to production without depending on it? 

Ejecting - as it is currently described - would only let me do it once. From technical perspective, there is no difference here - I could eject inside the CI and deploy that ejected server instead. So the difference only boils down to how we teach this.

Potential alternative is reaming `eject` to `build` command, which builds the server file (or maybe `prebuild`, to align with Expo).

**Comment by grabbou** [comment]
BTW! Start command could also open iOS/Android app via URL automatically once all models were build. The URL should open connection directly, without adding it. The connection should not be saved or duplicated. Users should be able to manually "save it" by tapping the Rise icon.
